### PR TITLE
Better array dtypes & support for array targets and T+N categorical forecasting

### PR DESCRIPTION
**Pull request by paxcema**
This PR is part refactor, part feature.

- Split `array` dtype into `num_array` and `cat_array`. Turns out current usage is inconsistent, so this was a needed piece of refactoring.
- Split `tsarray` dtype into `num_tsarray` and `cat_tsarray` for columns with single-valued rows that will be transformed into arrays because they are a time series.
- As part of the above, introduced array and categorical T+N time series target support with new encoders, accuracy functions, and updated mixer dispatch.
- Add tests to check new behavior. In particular, `test_array` shows (exciting, imho) performance where lightwood can easily learn to e.g. return "N-next numbers" or "reverse a sequence".

Benchmarks against `best` are [here](http://benchmarks.mindsdb.com:9107/compare/best/b134f0d5473ef753c27f863649d3dc40a39e7b90).

### CODE REVIEW

File: lightwood/analysis/helpers/acc_stats.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def analyze(self, info: Dict[str, object], **kwargs) -> Dict[str, object]:
     def fit(self, ns: SimpleNamespace, conf=Optional[np.ndarray]):
         self.col_stats = ns.dtype_dict
         self.target = ns.target
-        self.input_cols = [col for col in ns.dtype_dict.keys() if col != self.target]
+        self.input_cols = ns.input_cols
```
Comment (paxcema) [inline lightwood/analysis/helpers/acc_stats.py:33]: This is simpler and better than #832.

File: lightwood/analysis/nc/nc.py [hunk 2]
```diff
@@ -184,6 +184,21 @@ def apply_inverse(self, nc, significance):
         return np.vstack([nc[border], nc[border]])
 
 
+class ArrayBoostedAbsErrorErrFunc(BoostedAbsErrorErrFunc):
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/analysis/nc/nc.py:187]: Indeed. On second thought, I'll just remove it and keep the reference around in a gist or similar.

File: lightwood/data/cleaner.py [hunk 3]
```diff
@@ -191,6 +195,26 @@ def _standardize_array(element: object) -> Optional[Union[List[float], float]]:
...
+
+    Ex of edge case:
+    >> element = ['a']
+    >> _standardize_cat_array(element)
+    >> 'a'
+    """
+    if len(element) == 1:
+        return element[0]
+    else:
+        element = [e if e else '<UNK>' for e in element]
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/data/cleaner.py:213]: Noted, `bool(None)` evaluates to `False` so this clause is not needed anymore.

File: lightwood/encoder/array/array.py [hunk 4]
```diff
@@ -107,5 +107,25 @@ def decode(self, data: torch.Tensor) -> List[Iterable]:
         :param data: Encoded data prepared by this array encoder
         :returns: A list of iterable sequences in the original data space
         """
-        decoded = data.tolist()
+        decoded = self._normalizer.decode(data.tolist())
         return decoded
+
+
+class CatArrayEncoder(ArrayEncoder):
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/encoder/array/array.py:114]: We _could_ drop `ArrayEncoder`. The issue I see is that there would be quite a bit of code duplication between these two new encoders that inherit it. If you're okay with that, I'd be fine dropping it.

File: lightwood/encoder/array/ts_cat_array.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+from typing import List, Dict, Iterable, Optional
+
+import torch
+import torch.nn.functional as F
+
+from lightwood.encoder import BaseEncoder
+from lightwood.encoder.categorical import OneHotEncoder
+
+
+class TsCatArrayEncoder(BaseEncoder):
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/encoder/array/ts_cat_array.py:10]: I'll look into it, I don't want to litter that encoder with 100%-specific TS logic, but maybe there's a way.
